channelacceptor/rpcacceptor.go

```diff
@@ -0,0 +1,180 @@
+package channelacceptor
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+const (
+	// DefaultRPCTimeout is the default time that the Accept method will
+	// listen for responses from the RPC server, after which it will timeout.
+	DefaultRPCTimeout = 15 * time.Second
+)
+
+// OpenChannelResponse is a struct containing the response from the RPC client
+// which includes whether or not they accept the open channel request and to
+// which PendingChannelID the response is to.
+type OpenChannelResponse struct {
+	Accept           bool
+	PendingChannelID [32]byte
+}
+
+// ChanAcceptorClient represents an RPC streaming client that responds to
+// OpenChannel queries issued by the RPCAcceptor.
+type ChanAcceptorClient struct {
+	SendChan chan *OpenChannelRequest
+	RecvChan chan OpenChannelResponse
+
+	closeChan chan struct{}
+	active    bool
+}
+
+// RPCAcceptor represents the RPC-controlled variant of the ChannelAcceptor.
+// Only one RPC client is currently allowed at a time.
+type RPCAcceptor struct {
+	clientMtx sync.Mutex
+	client    *ChanAcceptorClient
+
+	acceptChan    chan *OpenChannelRequest
+	acceptClients map[[32]byte]chan bool
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// Accept is a predicate on the OpenChannelRequest which is sent to the RPC
+// client who will respond with the ultimate decision. This assumes a registered
+// client exists.
+func (r *RPCAcceptor) Accept(req *OpenChannelRequest) bool {
+	// Register a new channel for the PendingChannelID associated with this
+	// OpenChannelRequest.
+	responseChan := make(chan bool)
+	r.acceptClients[req.OpenChanMsg.PendingChannelID] = responseChan
+
+	select {
+	case r.acceptChan <- req:
+	case <-r.quit:
+		return false
+	}
+
+	for {
+		select {
+		case accept, _ := <-responseChan:
+			return accept
+		case <-time.After(DefaultRPCTimeout):
+			return false
+		case <-r.quit:
+			return false
+		}
+	}
+}
+
+// RegisterClient registers an RPC streaming client. There may only be one RPC
+// streaming client at a time.
+func (r *RPCAcceptor) RegisterClient() (*ChanAcceptorClient, error) {
+	r.clientMtx.Lock()
+	defer r.clientMtx.Unlock()
+
+	if r.active() {
+		return nil, fmt.Errorf("client already exists")
+	}
+
+	r.client = &ChanAcceptorClient{
+		SendChan:  make(chan *OpenChannelRequest),
+		RecvChan:  make(chan OpenChannelResponse),
+		closeChan: make(chan struct{}),
+		active:    true,
+	}
+
+	return r.client, nil
+}
+
+// UnregisterClient unregisters the RPC streaming client.
+func (r *RPCAcceptor) UnregisterClient() {
```

Comment:
Shouldn't this be on the `ChanAcceptorClient` struct? Otherwise, we can't individually unregister clients.

Reply:
Nvm, I see how the `RegisterClient` method works now, no strong preference, but it could also be done with an explicit constructor that instead passes in the `ChanAcceptorClient`.